File: server/lib/game_machine/game_systems/entity_tracking.rb
module GameMachine
  module GameSystems
    # @note Handles player location tracking and neighbor requests.
    #   The client must explicitly send a TrackEntit component to the server
    #   to be tracked, it doesn't happen automatically.
    #
    # @aspects TrackEntity
    # @aspects GetNeighbors
    class EntityTracking < Actor::Base
      include GameMachine::Commands

      aspect %w(TrackEntity)
      aspect %w(GetNeighbors)



      EXTRA = java.util.concurrent.ConcurrentHashMap.new
      attr_reader :grid, :extra_params, :aoe_grid

      def post_init
        if handler_klass = Application.config.entity_tracking_handler
         @tracking_handler = handler_klass.constantize.new
        end
        @entity_updates = []
        @grid = Grid.find_or_create('default')
        @aoe_grid = Grid.find_or_create('aoe')
        @paths = {}
        @width = grid.get_width
        @cell_count = grid.get_cell_count
        commands.misc.client_manager_register(self.class.name)
      end

      def on_receive(message)
        if message.is_a?(MessageLib::Entity)
          if message.get_neighbors
            send_neighbors(message)
          end

          # If a tracking handler is defined, it must return true to have
          # the player location saved
          if message.track_entity
            #GameMachine.logger.info "#{message.player.id} tracked"
            if @tracking_handler && message.entity_type == 'player'
              if @tracking_handler.verify(message)
                set_entity_location(message)
              end
            else
              set_entity_location(message)
            end
          end
        elsif message.is_a?(MessageLib::ClientManagerEvent)
          if message.event == 'disconnected'
            @grid.remove(message.player_id)
            @aoe_grid.remove(message.player_id)
            EXTRA.delete(message.player_id)
            GameMachine.logger.info "#{message.player_id} removed from grid"
          end
        else
          unhandled(message)
        end
      end

      private

      def set_entity_location(entity)
        vector = entity.vector3
        @grid.set(entity.id,vector.x,vector.y,vector.z,entity.entity_type)
        @aoe_grid.set(entity.id,vector.x,vector.y,vector.z,entity.entity_type)
        if entity.track_entity.has_track_extra
          track_extra = entity.track_entity.track_extra
          EXTRA[entity.id] = track_extra
        end
      end

      def location_entity(grid_value)
        entity = MessageLib::Entity.new.set_id(grid_value.id)

        entity.set_vector3(
          MessageLib::Vector3.new.
          set_x(grid_value.x.round(4)).
          set_y(grid_value.y.round(4)).
          set_z(grid_value.z.round(4))
        )

        if EXTRA.has_key?(grid_value.id)
          track_extra = EXTRA.fetch(grid_value.id)
          entity.set_track_extra(track_extra)
        end
        entity
      end

      def send_neighbors(message)
        type = message.get_neighbors.neighbor_type
        x = message.get_neighbors.vector3.x
        y = message.get_neighbors.vector3.y

        # Either .net or java protobufs have a bug with 0 floats
        if x.nil?
          x = 0
        end
        if y.nil?
          y = 0
        end
        search_results = grid.neighbors(x,y,type)
       
        neighbors = search_results.map do |grid_value|
          location_entity(grid_value)
        end
 
        if neighbors.empty?
          return
        end

        if message.has_player
          send_neighbors_to_player(neighbors,message.player)
        end
      end

      def send_neighbors_to_player(neighbors,player)
        neighbors.each_slice(30) do |group|
          entity = MessageLib::Entity.new.set_neighbors(
            MessageLib::Neighbors.new.set_entity_list(group)
          )
          entity.set_id(player.id)
          entity.set_player(player)
          entity.set_send_to_player(true)
          commands.player.send_message(entity,player.id)
        end
      end

    end
  end
end

File: server/java/src/main/java/com/game_machine/core/EntityTracking.java
package com.game_machine.core;

import java.util.ArrayList;

import GameMachine.Messages.ClientManagerEvent;
import GameMachine.Messages.Entity;
import GameMachine.Messages.Neighbors;
import GameMachine.Messages.Player;
import GameMachine.Messages.TrackData;
import akka.actor.ActorSelection;
import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;

public class EntityTracking extends UntypedActor {

	LoggingAdapter log = Logging.getLogger(getContext().system(), this);
	public static String name = "fastpath_entity_tracking";

	private Grid aoeGrid;
	private Grid grid;
	private ActorSelection messageGateway;

	public EntityTracking() {
		messageGateway = ActorUtil.getSelectionByName(MessageGateway.name);
		grid = Grid.find("default");
		aoeGrid = Grid.find("aoe");
		Commands.clientManagerRegister(name);
	}

	@Override
	public void onReceive(Object message) throws Exception {

		if (message instanceof Entity) {
			Entity entity = (Entity) message;

			SendNeighbors(entity);
			setEntityLocation(entity.trackData);

		} else if (message instanceof ClientManagerEvent) {
			ClientManagerEvent event = (ClientManagerEvent) message;
			if (event.event.equals("disconnected")) {
				removePlayerData(event);
			}
		} else {
			unhandled(message);
		}

	}

	private void removePlayerData(ClientManagerEvent event) {
		grid.remove(event.player_id);
		aoeGrid.remove(event.player_id);
	}

	private void SendNeighbors(Entity entity) {
		Float x = entity.trackData.x;
		Float y = entity.trackData.y;
		if (x == null) {
			x = 0f;
		}
		if (y == null) {
			y = 0f;
		}

		ArrayList<TrackData> trackDatas = grid.neighbors(x, y,
				entity.trackData.neighborEntityType);

		if (trackDatas.size() >= 1) {
			toNeighbors(entity.player, trackDatas);
		}
	}

	private void SendToGateway(Player player, Neighbors neighbors) {
		Entity playerMessage = new Entity();
		playerMessage.setNeighbors(neighbors);
		playerMessage.setPlayer(player);
		playerMessage.setId(player.id);
		messageGateway.tell(playerMessage, getSelf());
	}

	private void toNeighbors(Player player, ArrayList<TrackData> trackDatas) {
		int count = 0;
		Neighbors neighbors = new Neighbors();

		for (TrackData trackData : trackDatas) {
			neighbors.addTrackData(trackData);

			count++;
			if (count >= 30) {
				SendToGateway(player, neighbors);
				count = 0;
				neighbors = new Neighbors();
			}
		}
		SendToGateway(player, neighbors);
	}

	private void setEntityLocation(TrackData trackData) {
		
		// So either protostuff or protobuf-net has a bug where 0 floats come
		// through as null
		// This is *really* annoying must track down
		if (trackData.x == null) {
			trackData.x = 0f;
		}
		if (trackData.y == null) {
			trackData.y = 0f;
		}
		if (trackData.z == null) {
			trackData.z = 0f;
		}

		grid.set(trackData);
		aoeGrid.set(trackData);
	}

}
